uvicorn/protocols/utils.py

```diff
@@ -54,3 +74,44 @@ def get_path_with_query_string(scope: WWWScope) -> str:
     if scope["query_string"]:
         path_with_query_string = "{}?{}".format(path_with_query_string, scope["query_string"].decode("ascii"))
     return path_with_query_string
+
+
+def get_tls_info(transport: asyncio.Transport) -> dict[object, object]:
+    ###
+    # server_cert: Unable to set from transport information
+    # client_cert_chain: Just the peercert, currently no access to the full cert chain
+    # client_cert_name:
+    # client_cert_error: No access to this
+    # tls_version:
+    # cipher_suite: Too hard to convert without direct access to openssl
+    ###
+
+    ssl_info: dict[object, object] = {
+        "server_cert": None,
+        "client_cert_chain": [],
+        "client_cert_name": None,
+        "client_cert_error": None,
+        "tls_version": None,
+        "cipher_suite": None,
+    }
+
+    ssl_object = transport.get_extra_info("ssl_object", default=None)
+    peercert = ssl_object.getpeercert()
+
+    if peercert:
+        rdn_strings = []
+        for rdn in peercert["subject"]:
+            rdn_strings.append(
+                "+".join(
+                    [f"{RDNS_MAPPING[entry[0]]} = {entry[1]}" for entry in reversed(rdn) if entry[0] in RDNS_MAPPING]
+                )
+            )
+
+        ssl_info["client_cert_chain"] = [ssl.DER_cert_to_PEM_cert(ssl_object.getpeercert(binary_form=True))]
+        ssl_info["client_cert_name"] = ", ".join(rdn_strings) if rdn_strings else ""
+        ssl_info["tls_version"] = (
+            TLS_VERSION_MAP[ssl_object.version()] if ssl_object.version() in TLS_VERSION_MAP else None
+        )
+        ssl_info["cipher_suite"] = list(ssl_object.cipher())
```

Comment:
`cipher_suite` should be an integer.

Reply:
The issue here again lies with the specification. I would need a lookup table for every single cipher suite for TLSv1 through TLSv1.3.  The RFC referenced in the specification only has the ciphers for TLSv1.3.

This is another area where the specification doesn't easily align with what is readily available via the Python TLS module.

So rather than trying to collect and maintain a list of cipher mappings I opted for just passing through the result from the underlying SSLObject. This is obviously "off-spec" however. If we want to maintain strict compatibility with the spec then in a similiar regard to the `client_cert_name` we can just not set this value.

Reply:
Generate cipher suite values via something like https://gist.github.com/jschlyter/555438387a502f08ed05724b56db60ad perhaps?

Reply:
Pretty sure making a network request is a non-starter. This would immediately break for applications running in an offline/airgapped environment.

Reply:
> cipher_suite (integer or None) -- The TLS cipher suite that is being used. This is a 16-bit unsigned integer that encodes the pair of 8-bit integers specified in the relevant RFC, in network byte order. For example [RFC8446 section B.4](https://tools.ietf.org/html/rfc8446#appendix-B.4) defines that the cipher suite TLS_AES_128_GCM_SHA256 is {0x13, 0x01}; that is encoded as a cipher_suite value of 0x1301 (equal to 4865 decimal). Some web server implementations may be unable to provide this (e.g. if TLS is terminated by a separate proxy or load balancer); in that case set to None. Mandatory.

From the spec, the way I read this is that a vale is required but we can just supply None.

Reply:
> Pretty sure making a network request is a non-starter. This would immediately break for applications running in an offline/airgapped environment.

I'm not saying you should make a network request at runtime (that would be absurd…), but as part of a generated file when building uvicorn.

Reply:
Fetching that XML document each build tingles my spider senses. Even though we should normally be able to trust `https://www.iana.org` it would mean that something not reviewed and tested would be included in the builds.

Reply:
No need to fetch at each build, but a generated file with ciphers to code mapping is needed IMHO. How often it is regenerated and committed can be discussed. I would start adding the generated file to the repo, with a make target to rebuild on request. The list of ciphers is not updated that frequently.

Reply:
@mdgilene do you want a PR for your code adding `cipher_suite` support?

Reply:
@mdgilene https://github.com/jschlyter/uvicorn/tree/feature-asgiref-tls-extension-ciphers for cherry picking

Reply:
Pulled in the commit with some minor changes.